File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isValidBST(root) {
        if (!root) return true;
        return this.isValidBSTRec(root, root.left) && this.isValidBSTRec(root, root.right);
    }

    isValidBSTRec(root, node) {
        if (!node) return true;

        if (!this.findNode(root, node)) return false;

        return this.isValidBSTRec(root, node.left) && this.isValidBSTRec(root, node.right);
    }

    /**
     * @param {TreeNode} root
     * @param {TreeNode} node
     * @return {boolean}
     */
    findNode(root, node) {
        if (!root) return false;

        if (root == node) return true;

        if (root.val > node.val) return this.findNode(root.left, node);
        if (root.val < node.val) return this.findNode(root.right, node);

        return false;
    }
}

File: Data Structures & Algorithms/same-binary-tree/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isSameTree(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;
        if (p == null || q == null) return false;

        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.offer(p);
        queue.offer(q);

        while (!queue.isEmpty()) {
            p = queue.poll();
            q = queue.poll();

            if (p.val != q.val) return false;

            if (p.left != null && q.left != null) {
                queue.offer(p.left);
                queue.offer(q.left);
            } else if (!(p.left == null && q.left == null)) return false;

            if (p.right != null && q.right != null) {
                queue.offer(p.right);
                queue.offer(q.right);
            } else if (!(p.right == null && q.right == null)) return false;

        }

        return true;
    }
}
